File: MensLacrosseStats/View Controllers/Game/INSOGameDetailViewController.m
//
//  INSOGameDetailViewController.m
//  ScorebookLite
//
//  Created by James Dabrowski on 9/26/15.
//  Copyright © 2015 Intangible Software. All rights reserved.
//

#import "MensLacrosseStatsAppDelegate.h"

#import "INSOGameDetailViewController.h"
#import "INSOGameEditViewController.h"
#import "INSORosterPlayerSelectorViewController.h"
#import "INSOMensLacrosseStatsEnum.h"
#import "INSOGameStatsViewController.h"
#import "INSOEmailStatsViewController.h"

#import "Game.h"
#import "GameEvent.h"
#import "Event.h"
#import "RosterPlayer.h"

static NSString * INSOEditGameSegueIdentifier          = @"EditGameSegue";
static NSString * INSORecordStatsSegueIdentifier       = @"RecordStatsSegue";
static NSString * INSOGameStatsSegueIdentifier         = @"GameStatsSegue";
static NSString * INSOExportStatsSegueIdentifier       = @"ExportStatsSegue";


@interface INSOGameDetailViewController ()

// IBOutlets
@property (nonatomic, weak) IBOutlet UILabel* gameDateTimeLabel;
@property (nonatomic, weak) IBOutlet UILabel* homeTeamLabel;
@property (nonatomic, weak) IBOutlet UILabel* homeScoreLabel;
@property (nonatomic, weak) IBOutlet UILabel* visitingTeamLabel;
@property (nonatomic, weak) IBOutlet UILabel* visitingScoreLabel;
@property (nonatomic, weak) IBOutlet UILabel* locationLabel;

// IBActions
- (IBAction)editGame:(id)sender;
- (IBAction)recordStats:(id)sender;
- (IBAction)exportStats:(id)sender;

// Private Properties
@property (nonatomic) NSManagedObjectContext* managedObjectContext;
@property (nonatomic) NSInteger teamWatchingGoals;
@property (nonatomic) NSInteger otherTeamGoals;

// Private Methods

@end

@implementation INSOGameDetailViewController
#pragma mark - Lifecycle
- (void)viewDidLoad
{
    [super viewDidLoad];
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    
    [self configureView];
}

- (void)viewWillDisappear:(BOOL)animated
{
    NSError* error = nil;
    [self.managedObjectContext save:&error];
    if (error) {
        NSLog(@"Error saving context after a game: %@", error.localizedDescription); 
    }
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
}

#pragma mark - IBActions
- (void)recordStats:(id)sender
{
    [self performSegueWithIdentifier:INSORecordStatsSegueIdentifier sender:self];
}

- (void)editGame:(id)sender
{
    [self performSegueWithIdentifier:INSOEditGameSegueIdentifier sender:self];
}

- (void)exportStats:(id)sender
{
    [self performSegueWithIdentifier:INSOExportStatsSegueIdentifier sender:self];
}

#pragma mark - Private Properties
- (NSManagedObjectContext*)managedObjectContext
{
    if (!_managedObjectContext) {
        MensLacrosseStatsAppDelegate* appDelegate = (MensLacrosseStatsAppDelegate *)[[UIApplication sharedApplication] delegate];
        _managedObjectContext = appDelegate.managedObjectContext;
    }
    return _managedObjectContext;
}

#pragma mark - Private Methods
- (void)configureView
{
    NSString* dateFormat = [NSDateFormatter dateFormatFromTemplate:@"Mdyy" options:0 locale:[NSLocale currentLocale]];
    NSString* timeFormat = [NSDateFormatter dateFormatFromTemplate:@"hmma" options:0 locale:[NSLocale currentLocale]];
    NSString* dateTimeFormat = [NSString stringWithFormat:@"%@' at '%@", dateFormat, timeFormat];
    
    NSDateFormatter* formatter = [[NSDateFormatter alloc] init];
    [formatter setDateFormat:dateTimeFormat];
    self.gameDateTimeLabel.text = [formatter stringFromDate:self.game.gameDateTime];
    
    self.homeTeamLabel.text = self.game.homeTeam;
    self.visitingTeamLabel.text = self.game.visitingTeam;

    // Make sure the game object updates its own scores
    [self.game updateScores];
    
    self.homeScoreLabel.text = [NSString stringWithFormat:@"%@", self.game.homeScore];
    self.visitingScoreLabel.text = [NSString stringWithFormat:@"%@", self.game.visitorScore];
    
    self.locationLabel.text = self.game.location;
}

#pragma mark - Navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    if ([segue.identifier isEqualToString:INSOEditGameSegueIdentifier]) {
        [self prepareForGameEditSegue:segue sender:sender];
    }
    
    if ([segue.identifier isEqualToString:INSORecordStatsSegueIdentifier]) {
        [self prepareForRecordStatsSegue:segue sender:sender]; 
    }
    
    if ([segue.identifier isEqualToString:INSOGameStatsSegueIdentifier]) {
        [self prepareForGameStatsSegue:segue sender:sender]; 
    }
    
    if ([segue.identifier isEqualToString:INSOExportStatsSegueIdentifier]) {
        [self prepareForExportStatsSegue:segue sender:sender];
    }
}

- (void)prepareForGameEditSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    INSOGameEditViewController* dest = segue.destinationViewController;
    dest.game = self.game; 
}

- (void)prepareForRecordStatsSegue:(UIStoryboardSegue*)segue sender:(id)sender
{
    INSORosterPlayerSelectorViewController* dest = segue.destinationViewController;
    dest.game = self.game; 
}

- (void)prepareForGameStatsSegue:(UIStoryboardSegue*)segue sender:(id)sender
{
    INSOGameStatsViewController* dest = segue.destinationViewController;
    dest.game = self.game;
}

- (void)prepareForExportStatsSegue:(UIStoryboardSegue*)segue sender:(id)sender
{
    UINavigationController* navigationController = segue.destinationViewController;
    INSOEmailStatsViewController* emailViewController = [navigationController.viewControllers firstObject];
    emailViewController.game = self.game;
}


@end

File: MensLacrosseStats/View Controllers/Game/INSOGameDetailViewController.swift
//  Converted to Swift 5.5 by Swiftify v5.5.27463 - https://swiftify.com/
//
//  INSOGameDetailViewController.swift
//  ScorebookLite
//
//  Created by James Dabrowski on 9/26/15.
//  Copyright © 2015 Intangible Software. All rights reserved.
//

import UIKit
import SwiftUI

private var INSOEditGameSegueIdentifier = "EditGameSegue"
private var INSORecordStatsSegueIdentifier = "RecordStatsSegue"
private var INSOGameStatsSegueIdentifier = "GameStatsSegue"
private var INSOExportStatsSegueIdentifier = "ExportStatsSegue"

class INSOGameDetailViewController: UIViewController {
    // Public Properties
    var game: Game?

    // Private Properties
    private var managedObjectContext: NSManagedObjectContext {
        let appDelegate = UIApplication.shared.delegate as! AppDelegate
        return appDelegate.managedObjectContext
    }
    private var teamWatchingGoals = 0
    private var otherTeamGoals = 0
   
    // MARK: - Lifecycle

    override func viewDidLoad() {
        super.viewDidLoad()
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)

        configureView()
    }

    override func viewWillDisappear(_ animated: Bool) {
        do {
            try managedObjectContext.save()
        } catch {
            print("Error saving context after a game: \(error.localizedDescription)")
        }
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }

    // MARK: - IBOutlets
    @IBOutlet private weak var gameDateTimeLabel: UILabel!
    @IBOutlet private weak var homeTeamLabel: UILabel!
    @IBOutlet private weak var homeScoreLabel: UILabel!
    @IBOutlet private weak var visitingTeamLabel: UILabel!
    @IBOutlet private weak var visitingScoreLabel: UILabel!
    @IBOutlet private weak var locationLabel: UILabel!

    // MARK: - IBActions
    @IBAction func recordStats(_ sender: Any?) {
        performSegue(withIdentifier: INSORecordStatsSegueIdentifier, sender: self)
    }

    @IBAction func editGame(_ sender: Any?) {
        performSegue(withIdentifier: INSOEditGameSegueIdentifier, sender: self)
    }
    
    @IBSegueAction func exportStats(_ coder: NSCoder) -> UIViewController? {
        guard let game = game else {
            // If we don't have a game, we aren't going to segue.
            // Shouldn't actually have that export button enabled,
            // but that's a problem for another day. 
            return nil
        }
        return UIHostingController(coder: coder, rootView: ExportStatsView(fileGenerator: StatsFileGenerator(with: game)))
    }

    // MARK: - Private Methods
    func configureView() {
        let dateFormat = DateFormatter.dateFormat(fromTemplate: "Mdyy", options: 0, locale: NSLocale.current)
        let timeFormat = DateFormatter.dateFormat(fromTemplate: "hmma", options: 0, locale: NSLocale.current)
        let dateTimeFormat = "\(dateFormat ?? "")' at '\(timeFormat ?? "")"

        let formatter = DateFormatter()
        formatter.dateFormat = dateTimeFormat
        if let gameDateTime = game?.gameDateTime {
            gameDateTimeLabel.text = formatter.string(from: gameDateTime)
        }

        homeTeamLabel.text = game?.homeTeam
        visitingTeamLabel.text = game?.visitingTeam

        // Make sure the game object updates its own scores
        game?.updateScores()

        if let homeScore = game?.homeScore {
            homeScoreLabel.text = "\(homeScore)"
        }
        if let visitorScore = game?.visitorScore {
            visitingScoreLabel.text = "\(visitorScore)"
        }

        locationLabel.text = game?.location
    }

    // MARK: - Navigation
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if segue.identifier == INSOEditGameSegueIdentifier {
            prepare(forGameEdit: segue, sender: sender)
        }

        if segue.identifier == INSORecordStatsSegueIdentifier {
            prepare(forRecordStatsSegue: segue, sender: sender)
        }

        if segue.identifier == INSOGameStatsSegueIdentifier {
            prepare(forGameStatsSegue: segue, sender: sender)
        }
    }

    func prepare(forGameEdit segue: UIStoryboardSegue, sender: Any?) {
        let dest = segue.destination as? INSOGameEditViewController
        dest?.game = game
    }

    func prepare(forRecordStatsSegue segue: UIStoryboardSegue, sender: Any?) {
        let dest = segue.destination as? INSORosterPlayerSelectorViewController
        dest?.game = game
    }

    func prepare(forGameStatsSegue segue: UIStoryboardSegue, sender: Any?) {
        if let dest = segue.destination as? INSOGameStatsViewController {
            dest.game = game
        } 
    }
}
